WordPress/Classes/ViewRelated/Menus/Views/MenuItemCategoriesViewController.m

```diff
@@ -0,0 +1,156 @@
+                                   success:^(NSArray<PostCategory *> *categories) {
+                                       stopLoading();
+                                   } failure:^(NSError *error) {
+                                       stopLoading();
+                                       [self showLoadingErrorMessageForResults];
+                                   }];
+}
+
+#pragma mark - TableView methods
+
+- (void)configureSourceCell:(MenuItemSourceCell *)cell forIndexPath:(NSIndexPath *)indexPath
+{
+    PostCategory *category = [self.displayCategories objectAtIndex:indexPath.row];
+    
+    if ([self itemTypeMatchesSourceItemType] && [self.item.contentID integerValue] == [category.categoryID integerValue]) {
+        cell.sourceSelected = YES;
+    } else {
+        cell.sourceSelected = NO;
+    }
+    
+    [cell setTitle:category.categoryName];
+    
+    NSInteger indentationLevel = [[self.categoryIndentationDict objectForKey:[category.categoryID stringValue]] integerValue];
+    [cell setSourceHierarchyIndentation:indentationLevel];
+}
+
+- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    [super tableView:tableView didSelectRowAtIndexPath:indexPath];
+    
+    PostCategory *category = [self.displayCategories objectAtIndex:indexPath.row];
+    [self setItemSourceWithContentID:category.categoryID name:category.categoryName];
+    
+    [self deselectVisibleSourceCellsIfNeeded];
+    
+    MenuItemSourceCell *selectedCell = [tableView cellForRowAtIndexPath:indexPath];
+    selectedCell.sourceSelected = YES;
+}
+
+#pragma mark -
+
+- (void)updateDisplayCategories
+{
+    // Get sorted categories by parent/child relationship
+    self.categoryIndentationDict = [NSMutableDictionary dictionary];
+    
+    // Get sorted categories by parent/child relationship
+    WPCategoryTree *tree = [[WPCategoryTree alloc] initWithParent:nil];
+    [tree getChildrenFromObjects:self.resultsController.fetchedObjects];
+    
+    self.displayCategories = [tree getAllObjects];
+    
+    // Get the indentation level of each category.
+    NSMutableDictionary *categoryDict = [NSMutableDictionary dictionary];
+    for (PostCategory *category in self.displayCategories) {
+        [categoryDict setObject:category forKey:category.categoryID];
+    }
+    for (PostCategory *category in self.displayCategories) {
+        NSInteger indentationLevel = [self indentationLevelForCategory:category.parentID
+                                                    categoryCollection:categoryDict];
+        [self.categoryIndentationDict setValue:[NSNumber numberWithInteger:indentationLevel]
+                                        forKey:[category.categoryID stringValue]];
+    }
+    
+    [self.tableView reloadData];
+}
+
+- (NSInteger)indentationLevelForCategory:(NSNumber *)parentID categoryCollection:(NSMutableDictionary *)categoryDict
+{
+    if ([parentID intValue] == 0) {
```

Comment:
Maybe out of scope but we should really think about storing `depth` in the `PostCategory` model rather than recompute it all the time.